Sources/SourceKitLSP/Swift/SyntacticTestIndex.swift

```diff
@@ -0,0 +1,229 @@
+      return !lhsUris.intersection(rhsUris).isEmpty
+    }
+  }
+}
+
+/// Data from a syntactic scan of a source file for tests.
+fileprivate struct IndexedTests {
+  /// The tests within the source file.
+  let tests: [TestItem]
+
+  /// The modification date of the source file when it was scanned. A file won't get re-scanned if its modification date
+  /// is older or the same as this date.
+  let sourceFileModificationDate: Date
+}
+
+/// Syntactically scans the file at the given URL for tests declared within it.
+///
+/// Does not write the results to the index.
+///
+/// The order of the returned tests is not defined. The results should be sorted before being returned to the editor.
+fileprivate func testItems(in url: URL) async -> [TestItem] {
+  guard url.pathExtension == "swift" else {
+    return []
+  }
+  let syntaxTreeManager = SyntaxTreeManager()
+  let snapshot = orLog("Getting document snapshot for swift-testing scanning") {
+    try DocumentSnapshot(withContentsFromDisk: url, language: .swift)
+  }
+  guard let snapshot else {
+    return []
+  }
+  async let swiftTestingTests = SyntacticSwiftTestingTestScanner.findTestSymbols(
+    in: snapshot,
+    syntaxTreeManager: syntaxTreeManager
+  )
+  async let xcTests = SyntacticSwiftXCTestScanner.findTestSymbols(in: snapshot, syntaxTreeManager: syntaxTreeManager)
+  return await swiftTestingTests + xcTests
+}
+
+actor SyntacticTestIndex {
+  /// The tests discovered by the index.
+  private var indexedTests: [DocumentURI: IndexedTests] = [:]
+
+  /// Files that have been removed using `removeFileForIndex`.
+  ///
+  /// We need to keep track of these files because when the files get removed, there might be an in-progress indexing
+  /// operation running for that file. We need to ensure that this indexing operation doesn't write add the removed file
+  /// back to `indexTests`.
+  private var removedFiles: Set<DocumentURI> = []
+
+  /// The queue on which the index is being updated and queried.
+  ///
+  /// Tracking dependencies between tasks within this queue allows us to start indexing tasks in parallel with low
+  /// priority and elevate their priority once a read task comes in, which has higher priority and depends on the
+  /// indexing tasks to finish.
+  private let indexingQueue = AsyncQueue<TaskMetadata>()
+
+  init() {}
+
+  private func removeFilesFromIndex(_ removedFiles: Set<DocumentURI>) {
+    self.removedFiles.formUnion(removedFiles)
+    for removedFile in removedFiles {
+      self.indexedTests[removedFile] = nil
+    }
+  }
+
+  /// Called when the list of files that may contain tests is updated.
+  ///
+  /// All files that are not in the new list of test files will be removed from the index.
+  func listOfTestFilesDidChange(_ testFiles: [DocumentURI]) {
+    let removedFiles = Set(self.indexedTests.keys).subtracting(testFiles)
+    removeFilesFromIndex(removedFiles)
+
+    rescanFiles(testFiles)
```

Comment:
Should rescanning reset removed files? If a file was re-added (say switching branches), seems like we'd forever ignore it right now?

Reply:
Oh, very good catch. Fixed it and added a test for it.